### PR TITLE
Move the EnvoyCarrier proto out of the C++ tracer repo

### PR DESCRIPTION
**Pull request by jmacd**

### CODE REVIEW

File: envoy_carrier.proto [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+package lightstep.envoy;
+
+// CarrierStruct stores a set of key-values containing an OpenTracing span
+// context, as carried by Envoy, suitably encoded as a serialized protocol
+// buffer for transport via an HTTP header.
+//
+// The OpenTracing basictracer specifies the following keys and values:
+//
+//   "ot-tracer-traceid": 64 bit trace identifier encoded in 16 hex bytes
```
Comment (bhs) [inline envoy_carrier.proto:11]: just mentioning: I am enthusiastic about moving to a single key for the trace+span+sampled, and a single key for all baggage. but I don't want to take on that migration right now, so "just mentioning".

**Review by bhs — COMMENTED** [review]
@jmacd the other obvious thing would be to represent the various fields as native proto rather than as a generic map. Is there a short version for why we didn't do that?

**Comment by jmacd** [comment]
The short version is that I used the existing CarrierReader and CarrierWriter implementations, which arrange call Set(...) and Foreach(...) with these keys. It's not difficult to change this, and now would be the time to do it.

**Comment by bhs** [comment]
@jmacd it's a tough call... if this is really about Envoy, it's not properly our decision to make (i.e., what they want to be able to support).

Let's stick with what you have since we know it works. I am tempted to pick the "perfect" format, but in reality that should be done out in the open and could take months if there's a real effort to get consensus (which I think is warranted here). There's already the text-based propagation format, so this would be the binary sibling.

**Comment by jmacd** [comment]
I'm going to presume you've thought through this more than me, so I'll carry on. I worry about the amount of work that will be thrown away if we implement this format and then decide to do another down the road.

**Comment by jmacd** [comment]
I will merge this and point the Envoy folks here for discussion.

**Pull request merged**